### PR TITLE
allow to erase other than K64F devices

### PR DESCRIPTION
**Pull request by jupe**
## Status
**READY**

## Description
allow to erase other than K64F devices, when failing return meanful return code.

## Todo
- [x]  Test
- [ ]  Documentation (follow-up PR - codes need to consolidate first)

### CODE REVIEW

File: test/test_erase.py [hunk 1]
```diff
@@ -87,6 +86,38 @@ def test_erase_with_target_id(self):
...
 
+    # test func name is larger than 30, but is meaningful
+    # pylint: disable=invalid-name
+    def test_erase_failed_non_supported_devices(self):
+        mbeds = mbed_lstools.create()
+        devices = mbeds.list_mbeds(read_details_txt=True)
+        devices = [dev for dev in devices if dev.get('daplink_automation_allowed') == '0']
+        count = len(devices)
+        if count == 0:
+            self.skipTest('no suitable test')
```
Comment (jupe) [inline test/test_erase.py]: This is workaround for https://github.com/ARMmbed/mbed-ls/issues/288

Comment (bridadan) [reply test/test_erase.py]: Sorry about that, we'll work on getting a patch up for that soon

Comment (jupe) [reply test/test_erase.py]: this is not big problem - at least in this case :)

File: mbed_flasher/erase.py [hunk 2]
```diff
@@ -145,8 +145,45 @@ def runner(self, mount_point, filename):
...
+        self.logger.info("method used for reset: %s", method)
+        available_devices = self.get_available_device_mapping()
+
+        if target_id is None:
+            return EXIT_CODE_TARGET_ID_MISSING
+
+        targets_to_erase = self.prepare_target_to_erase(target_id, available_devices)
+
+        if len(targets_to_erase) <= 0:
+            print("Could not map given target_id(s) to available devices")
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Prefer using logger.

File: mbed_flasher/erase.py [hunk 3]
```diff
@@ -145,8 +145,45 @@ def runner(self, mount_point, filename):
...
+        for item in targets_to_erase:
+            if method == 'simple' and 'mount_point' in item and 'serial_port' in item:
+                erase_fnc = self._erase_board_simple
+            elif method == 'pyocd':
+                erase_fnc = self._erase_board_with_pyocd
+            elif method == 'edbg':
+                self.logger.warning("Not supported yet")
+                return EXIT_CODE_IMPLEMENTATION_MISSING
+            else:
+                print("Selected method %s not supported" % method)
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Prefer using logger.

File: mbed_flasher/erase.py [hunk 4]
```diff
@@ -211,56 +248,24 @@ def erase_board(self, target, no_reset):
...
-            else:
-                print("Selected method %s not supported" % method)
-                return EXIT_CODE_NONSUPPORTED_METHOD_FOR_ERASE
-
+    def _erase_board_with_pyocd(self, target, no_reset):
+        try:
+            from pyOCD.board import MbedBoard
+            from pyOCD.pyDAPAccess import DAPAccessIntf
+        except ImportError:
+            print('pyOCD missing, install it\n')
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Prefer using logger.

File: mbed_flasher/erase.py [hunk 5]
```diff
@@ -211,56 +248,24 @@ def erase_board(self, target, no_reset):
...
+            return EXIT_CODE_PYOCD_MISSING
+        board = MbedBoard.chooseBoard(board_id=target["target_id"])
+        self.logger.info("erasing device")
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Could add the device identification for the print.

File: mbed_flasher/erase.py [hunk 6]
```diff
@@ -211,56 +248,24 @@ def erase_board(self, target, no_reset):
...
+        ocd_target = board.target
+        flash = ocd_target.flash
+        try:
+            flash.eraseAll()
+            if not no_reset:
+                ocd_target.reset()
+        except DAPAccessIntf.TransferFaultError:
+            pass
+        self.logger.info("erase completed")
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Could add the device identification for the print.

File: mbed_flasher/erase.py [hunk 7]
```diff
@@ -145,8 +145,45 @@ def runner(self, mount_point, filename):
...
+            print("Could not map given target_id(s) to available devices")
+            return EXIT_CODE_COULD_NOT_MAP_TO_DEVICE
+
+        for item in targets_to_erase:
+            if method == 'simple' and 'mount_point' in item and 'serial_port' in item:
+                erase_fnc = self._erase_board_simple
+            elif method == 'pyocd':
+                erase_fnc = self._erase_board_with_pyocd
+            elif method == 'edbg':
+                self.logger.warning("Not supported yet")
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Use error level?

File: mbed_flasher/erase.py (hunk 2 above)
Comment (jupe) [reply mbed_flasher/erase.py]: true - better to change in this phase (originally was prints)

File: mbed_flasher/erase.py (hunk 7 above)
Comment (jupe) [reply mbed_flasher/erase.py]: true

**Comment by jupe** [comment]
@juhhov could you re-review this since there is multiple commits after last review :)

**Pull request merged**